pkg/server/delta/v3/server.go

```diff
@@ -210,17 +213,35 @@ func (s *server) DeltaStreamHandler(str stream.DeltaStream, typeURL string) erro
 }
 
 // When we subscribe, we just want to make the cache know we are subscribing to a resource.
-// Providing a name with an empty version is enough to make that happen.
-func (s *server) subscribe(resources []string, sv map[string]string) {
+// Even if the stream is wildcard, we keep the list of explicitly subscribed resources as the wildcard subscription can be discarded later on.
+func (s *server) subscribe(resources []string, streamState *stream.StreamState) {
+	sv := streamState.GetSubscribedResourceNames()
 	for _, resource := range resources {
-		sv[resource] = ""
+		if resource == "*" {
+			streamState.SetWildcard(true)
+			continue
+		}
+		sv[resource] = struct{}{}
 	}
 }
 
-// Unsubscriptions remove resources from the stream state to
-// indicate to the cache that we don't care about the resource anymore
-func (s *server) unsubscribe(resources []string, sv map[string]string) {
+// Unsubscriptions remove resources from the stream subscribed resources.
+// If explicitly unsubscribing from wildcard, the stream is updated and is now watching only subscribed resources.
+func (s *server) unsubscribe(resources []string, streamState *stream.StreamState) {
+	sv := streamState.GetSubscribedResourceNames()
 	for _, resource := range resources {
+		if resource == "*" {
+			streamState.SetWildcard(false)
+			continue
+		}
+		if _, ok := sv[resource]; ok && streamState.IsWildcard() {
+			// xds protocol specifically states that if a resource if unsubscribed while a wildcard watch is present,
```

Comment:
Can we clean this comment block up?

Reply:
I tried to make it clearer, though it is still so counter-intuitive to me I have a hard time formalizing it